Distinct kill and restart actions

SimonMarquis opened the pull request:
- kill: simply kills the process
- restart: kills the process and restarts it

Reorder the notification actions to Restart > Kill > Cancel.
This order feels more natural for what this library tries to achieve.

Fixes #20

![16-25-46 6712562](https://user-images.githubusercontent.com/1921278/172055445-544a46a0-e59d-4609-8acd-4cd3eca9bd8b.png)

SimonMarquis committed:
Distinct kill and restart actions

- kill: simply kills the process (`android:taskAffinity` is required for `Intent.FLAG_ACTIVITY_TASK_ON_HOME` to work as expected.)
- restart: kills the process and restarts it

Reorder the notification actions to Restart > Kill > Cancel.
This order feels more natural for what this library tries to achieve.
And rename `DeathActivity` to `VenomActivity` to host both kill & restart actions.

Fixes #20

SimonMarquis commented:
We could also update [preview/header.png](preview/header.png) and  [preview/preview.gif](preview/header.gif).

SimonMarquis commented:
Waiting for #35 to be merged.

YarikSOffice commented on venom/src/main/java/com/github/venom/service/VenomService.kt:
```diff
@@ -49,11 +50,17 @@ internal class VenomService : Service() {
 
+    private fun kill() {
+        stopSelf()
+        Process.killProcess(Process.myPid())
```
Hey Simon! Thanks for the contribution

According to my testing, the new button still restarts the app. Perhaps, I'm missing something - let's try to figure it out.

Have you run the sample app to test this solution?
It looks like you are using the same function to kill the app as we do for the kill-restart scenario. Any difference?

SimonMarquis replied:
My bad, I tested it once and assumed it worked. Unfortunately the OS tries to be "smart" and restart the process if it detects an Activity was not correctly stopped.

SimonMarquis commented:
But... I think I found a workaround that could be used to really simulate a kill (without a restart).

The trick is to start a new `Activity` with the [FLAG_ACTIVITY_TASK_ON_HOME](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_TASK_ON_HOME) and a "unique" [android:taskAffinity](https://developer.android.com/guide/topics/manifest/activity-element#aff).

This will have the effect of starting this new Activity "from" the launcher, and killing the process at this point will not re-start the whole application.

This can also be tested by starting the launcher and immediately killing the process, though it can bring the intent chooser:

```kotlin
val launcher = Intent(Intent.ACTION_MAIN).apply { 
    addCategory(Intent.CATEGORY_HOME)
    flags = Intent.FLAG_ACTIVITY_NEW_TASK
}
startActivity(launcher)
```


Here is what it looks like with an intermediate Activity:
As you can seen, the original activity stack is restored, and the process id has changed.

https://user-images.githubusercontent.com/1921278/173668610-69f9f629-b634-4b7a-a9ef-b9b40ee110bd.mp4

YarikSOffice commented:
@SimonMarquis could you please share a code snippet to achieve this behavior? 

I'm trying to replicate this and add finally this feature if your solution is stable enough.

SimonMarquis commented:
Hi @YarikSOffice, here is the updated version.
Please tell me if you can reproduce this behavior on your end.

Also, I've taken the liberty of renaming `DeathActivity` to `VenomActivity` since it can host both kill & restart actions without any code change.

SimonMarquis commented:
Hi Yaroslav, had some time to review this?
Can you reproduce it on your devices?
Tell me if you need more info.

YarikSOffice committed:
Merge branch 'master' into feat/kill-or-restart

YarikSOffice committed:
Fix the service declaration

SimonMarquis committed:
Update Android instrumented tests to click the restart button

YarikSOffice commented:
@SimonMarquis Hey! 
Let's try to stabilize this solution and maybe add new tests. If this doesn't work, I think there is another way to kill the process without triggering a restart.

SimonMarquis commented:
Hi @YarikSOffice,
I've found the issue. It was something I did not test during my expermientations.
I initially thought we could use the same Activity for both kill & restart actions, but it turns out `android:taskAffinity` will prevent the restart action to properly restart the Activity.
Well therefore have to split both Activities, but it can be done in a minimalist way.

SimonMarquis committed:
Split `VenomActivity` into two distinct `Kill` and `Restart`

to keep the `android:` properties in the AndroidManifest separated.
This had side-effects, leading to the restart action not restarting a "backgrounded" Activity.

YarikSOffice commented:
@SimonMarquis curious what is the reason for the activity not being restarted properly? in the case of task affinity

i didn't see any issues while making tests

SimonMarquis commented:
I think forcing a different task activity would lead to 2 tasks being created (1 with your app activity,  1 with the Venom Activity as a root) when the kill action is requested.
And I guess the system only tries to restat the most recent tasks, which is now empty. 🤷

SimonMarquis committed:
Fix typos

SimonMarquis commented:
I'd love to add tests for this action as well, but could you take a look at #37 before, and merge it if you are ok with these changes?

YarikSOffice committed:
Merge branch 'master' into feat/kill-or-restart

YarikSOffice commented:
I merged the PR with tests. but it seems like the build is failing now

SimonMarquis commented:
Really strange to be honest. Can you re-launch the checks?

SimonMarquis commented:
Also, since the new tests will be kind of copies of the previous tests, would you be ok if I migrate to [TestParameterInjector](https://github.com/google/TestParameterInjector)?

It would simplify the test matrix _a lot_ with a single test method.

YarikSOffice commented:
I'm currently testing another way to get it working

https://github.com/YarikSOffice/venom/pull/41

YarikSOffice commented:
[TestParameterInjector](https://github.com/google/TestParameterInjector) looks good to me

SimonMarquis committed:
Test the kill behavior

- Migrate test matrix to TestParameterInjector to avoid repetitions
- Testing for process death is done through PackageManager's runningAppProcesses

YarikSOffice commented:
Thanks for your contribution Simon!

I'm closing this PR in favor of https://github.com/YarikSOffice/venom/pull/41 which is heavily based on yours. Really appreciate your efforts 👍 
Eventually, I decided to avoid the test matrix since it complicates debugging if a specific test is failed.

Sorry for the late resolution of this PR.
